### ISSUE TITLE
User Validation Service

### ISSUE THREAD

**Issue by WorldMaker** [walkeriniraq/twitarr#263]
As a developer of another JoCoCruise ecosystem web app, I would like an easy to use User Validation Service so that my users/players/Seamonkey compatriots can use one Twit-Arr login of convenience.

Nathan and I discussed this a bit today. Chat transcript for helpfulness (reformatted quick and dirty from Facebook Messenger into some semblance of simple Markdown):

**Max:** Hey Nathan! Some questions regarding project "dinghy": It would make sense to share user db with Twit-arr, right? Would it just be a matter of expecting the right mongo collection? Any tips for working with it other than the obvious "read the twit-arr source"? If I need to add things to the User schema should I try to pr them back to twit-arr just in case they are useful? I guess there is an admin page in twit-arr they might be twiddled? Thanks!

**Nathan:** Yes, PR for user changes

I was thinking of putting a user validation service into Twitarr - we might have one already

if not it's pretty easy to do

Nvm - we don't have a validation service - I'll add one though

You may want your own user objects though - assume a username that comes from the user auth service

and then use that to key your user object off of

**Max:** I was thinking a simple service to send out tweets might be handy too. There might be a couple cases were it would be cool to tweet something out in case it brings in interest from non-players. Maybe just use something like the Twitter messaging intent screen.

**Nathan:** there is already an api for the CM app - could reuse most of that

**Max:** Cool. I'll take a look at that if we get to a point for that kind of a "stretch" feature.

**Nathan:** sounds good.  smile emoticon

**Max:** It would be cool if the user auth service passed along the avatar image path. Also, I guess I need to figure out how I would bootstrap DMs/Admins from the validation service information.

I guess it'd also be nice to have a simple test version of the validation service I can run without running the full Twit-Arr during development.

**Nathan:** I can send back the avatar path. I think twitarr admins may be separate from the DMs

I should be putting this into a ticket

**Max:** I'm sure Twitarr Admin and DMs will be separate.

**Nathan:**  So the way I've done this previously is to have you forward the users to a specially-formatted URL

**Max:** That makes sense. Use a nonce/capability URL.

**Nathan:** they'll log in with twit-arr credentials, and get redirected back to dinghy with a key appended to the url

you'll trade that key to the API for a user "packet" containing the username, avatar, and anything else you want

(I hijacked that scheme from oauth)  smile emoticon

**Max:** Yup, that sounds fair and familiar.

**Nathan:** So for test purposes, you can either assume a user in the dev code (i.e. fake it) or build a fake auth service that returns static data

**Max:** Looks like I'll get the excitement of trying to build a passport (node auth lib) plugin.

Although I guess if you could go pure/true OAuth 2 or OAuth Connect I maybe could piggyback existing tools.

As much fun as those are to implement. tongue emoticon

**Nathan:** Up to you - I didn't find it too difficult to build in rails - it was like 10 lines of code once I pulled in a  http library

**Max:**  Could be worth a shot to see if we could leverage standard OAuth.

You are right, it does seem like the libraries are increasingly standard, and I guess that even includes running an idP in Ruby.

(end scene)

**Comment by tasermonkey** [comment]
+1

**Comment by Wingar** [comment]
+1 but I'd rather look at alternatives to OAuth. I have a thing for OAuth... not a good thing.

**Comment by WorldMaker** [comment]
Yeah, I have mixed feelings about OAuth. On the one hand it's a standard and there are a bunch of handy libraries for it. On the other hand, yup it's not the greatest standard when it comes to things like simplicity or ease of debugging...

**Comment by Wingar** [comment]
Plus a lot of the advantages and purpose of OAuth is lost on us, really. OAuth is meant to be used in the internet, letting users access other services with other login credentials, while keeping those credentials hidden from said other services, as you all know.
Really... we're in a closed environment and we can generally trust others, hell we already do with CruiseMonkey. We can probably just get by with a standard login API.

**Comment by tasermonkey** [comment]
Well with that argument, I would say, we could probably just get by with a single http header: `USER_NAME: <username>` .  Though we do have a tech minded community, I feel like they are pretty trust worthy. :-P

**Comment by Wingar** [comment]
Yeah, I'm generally not a fan of ditching security but if we were to pre-hash the passwords before moving them across the network, we're really not taking any more of a risk than we are now.

**Comment by WorldMaker** [comment]
I'm kind of wondering how tough it would be to do a boat-locked browserid...

**Comment by Wingar** [comment]
We already have timed key system in place. Look at `/api/v2/user/auth`

**Comment by Wingar** [comment]
If we were to assume we had DNS next year, and it worked, we could just run them under the same vhost and use some kind of shared cookie. It at least covers browsers.

**Comment by Wingar** [comment]
That might a bit of an ambitious assumption though...

**Comment by tasermonkey** [comment]
Yeah, I would **not** hold my breath on a functioning DNS resolution.

**Comment by Wingar** [comment]
Actually, it would still work without DNS, if it's running under the same vhost and is just accessed by different addresses, like CruiseMonkey's stuff is/was.

**Comment by tasermonkey** [comment]
right. that ofcourse would work :-)

**Comment by WorldMaker** [comment]
I don't think it is easy to read a Rails auth session cookie from Node, though. Maybe if we cookied a JWT or something. Also, of course, cookies don't scale in the painful case if/when we need multiple magic numbers/vhosts as we grow the number of shipboard apps...

**Comment by Wingar** [comment]
We don't have to use any rails-tied stuff, it's pretty easy to freely write cookies as you want.

**Comment by Wingar** [comment]
What I was getting at would our rails session would just be that, and auth would be a seperate cookie for both apps.

**Comment by WorldMaker** [comment]
So here's what I'm currently thinking about for "BoatAuth" after more research than this topic probably deserves. Features that would be nice to have:

- Simple :exclamation: 
- *Single* Sign-On, where if a user is already logged in on Twit-Arr no need to sign in to second app
- Resilient Against DNS
- Resilient In Case of Scaling/Spreading Apps across Servers/Domains/IP Addresses
- At Least Somewhat Secure; we all are mostly friends and it should be unlikely for shenanigans, but it would be nice to at least avoid the easy "hacks"

I keep coming back to [JSON Web Tokens](http://jwt.io) as the best bet for underlying technology. It looks like there is a reasonable coverage in JWT libraries out there. This provides us at least a basic level of security in a somewhat standardized form that we can reuse existing code for.

Bootstrap:

1. Twit-Arr generates a BoatAuth public key/private key pair. Posts the public key somewhere usefully accessible to Ship App Devs, such as this here GitHub page. (Keeps the private key a secret, of course.) (Self-signed should be fine; probably no reason to get a fancy certificate.)
2. Decide if we want to `aud` audience checks and whitelists, and agree on URLs to use. (Luckily these don't actually have to be "live" DNS.)
  - Pro: Makes it a bit tougher to reuse JWTs
  - Con: Maybe "too secure" for BoatAuth needs; removes Cookie option
3. On boat, BoatAuth apps would need live Twit-Arr address, but hopefully not necessarily vice versa (depending maybe on `aud` whitelisting)

App Usage:

1. App checks `AUTH_TOKEN` cookie for JWT (if not `aud` checked)
2. App checks for `Authorization: JWT *token*` header
3. App checks for `auth_token` in query string and/or request body
4. If a token is found, decode/verify it with the Twit-Arr/BoatAuth Public Key. `sub` should of course be the Twit-Arr handle/username. We might want other BoatAuth-specific claims.
5. If not found, the App can redirect to a Twit-Arr page providing a return callback URL that can accept the `auth_token` in query string for the currently logged in user (logging in the user if not logged in)
6. It's up to the app to session store the `auth_token` as it sees fit (such as bootstrapping into a more interesting login session or continue to pass the `auth_token` around somehow)

It would be great if the same JWT could be passed back to Twit-Arr APIs, but maybe not required.

I think this meets the requirements:

- It's reasonably simple, and there are reusable bits of libraries out there for working with JWTs. It's not full OAuth; it's BoatAuth. (Bonus: same JWT can be used for sessions and stateless REST API calls/apps.)
- It supports single sign on for both same machine/magic number/dns (cookie) and any split/multi-server scenarios
- It's At Least Somewhat Secure in that all the Apps should be able to trust (with a good private key/public key pair) that Twit-Arr is the only thing creating the JWT tokens for logged in users. There's still ways to "steal" another user's JWT, but there's also a path to greater security *if need be/desired* in implementing `aud` checks, `exp` expiration policies, encouraging apps to limit when `auth_token` is passed by query string, etc. We probably wouldn't need to go down those paths, but it would be nice knowing they were there.

From my examining the NodeJS libraries I would be using, I think about the only thing that I wouldn't get "out of the box" is the check for a JWT cookie. Which seems like a good place to be, and some hope that maybe this scheme is workable...

Another thought, theoretically we should even be able to create simple test JWTs for logins under this scheme, using say JWTs expiring just before Go Live on the actual ship, that we could copy and paste to test without needing an actual Twit-Arr running...

**Comment by Wingar** [comment]
I can easily +1 all this, however I think having the `auth_token` query string is unnecessary, given the other two solutions should cover everything just fine. 

If everyone's happy with this, the next question is what data do we need to store in the token? Are we gonna go with our already-implemented timed API key system (The one behind `/api/v2/user/auth`) + Metadata (username, etc)?

**Comment by WorldMaker** [comment]
`auth_token` query string is useful for login redirect flow for single sign on when cookies can't be shared, say due to separate vhosts or what have you, example:

```
http://app1/some/page => http://twitarr/?return_url=http://app1/some/page => http://app1/some/page?auth_token=*JWT*
```

Where `=>` is plain old HTTP 303 redirect. You can't send an Authorization header in a 303 redirect and while there are ways to initiate a POST and include it in the request body, there are issues with cross-domain POSTs in browsers that are better off ignored.

**Comment by Wingar** [comment]
Okay then, that's fair enough, however I will say we probably won't be dealing with multiple DNS records or even IPs, still gotta live within the boats network. Should be pretty simple to implement and test out.

**Comment by WorldMaker** [comment]
Yup, it's a resiliency measure in case something crazy happens. Better to have it in place and test it, than to need it and not have it. I'm just picturing some crazy situation where an app is chewing CPU and RAM and has to get booted to another laptop with a different IP address or some weird case of say a device sending IPv4 address for one app and IPv6 address for another app. Again, stupid worst cases, but still useful to plan for.

**Comment by RangerRick** [comment]
FYI, 303 redirects do not play well with CORS, so I fear you may run into issues there.  From what I understand, a lot of ReST stuff has settled on setting a Location: header without the redirect and relying on clients to handle it.

**Comment by WorldMaker** [comment]
@RangerRick CORS is not an issue with 303 redirects when all the data is passed via query string, which is [the reason I suggested making sure query strings are supported by applications](https://github.com/walkeriniraq/twitarr/issues/263#issuecomment-112297582) as an `auth_token` vector so that *if* the worst should happen and we do need a 303 fallback, applications can still play ball. The whole point of the 303 fallback, as a fallback, in this plan is to avoid, prevent, and prepare for the CORS worst case. It's essentially the old fashioned "OpenID Classic" workflow: if all else fails do a full browser redirect (not in an AJAX call!) and hope maybe someday the user will return with a valid credential.

**Comment by WorldMaker** [comment]
Side thought: I wonder if it would be worth investigating IPv6 techniques for kinder to learn magic numbers this cruise. Most cellphones have to support IPv6 these days at least (AT&T and Verizon are apparently both near requiring it). Certainly `fe80::beef` would be faster to communicate than `168.127.13.10`. The fly in the ointment is likely whatever router is on the ship.

**Comment by RangerRick** [comment]
Ah, you're right.  Forgot that GET => GET is OK.

As for IPv6, knowing what those shipboard systems are like, I'd bet they'll be one of the last places in the world to support it.

**Comment by Wingar** [comment]
I'm late here (Starting to look at Twitarr stuff again :) ) but the shipboard network doesn't use v6, so that's very, very unlikely. I remember explicitly running a DHCPv6 request to see if it did, so I wouldn't count on that. Sure that's just DHCP, but I'm willing to bet the switching gear doesn't like v6. At the very least I wouldn't count on it, we can do some experimenting on JCC6 and see what we come up with for next time.

**Comment by RangerRick** [comment]
I had scarface ask some questions of the ship folks, and they don't support multicast nor v6, so most of our handy discovery ideas would probably not work.  :(
